transcom/mymove
# B21526-INT-add db swagger support for actual expense reimbursement

## Agility ticket
[B21526](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21526)

## Summary
The change was to update the db and swagger api to support Actual Expense Reimbursement PPM's.  This was achieved by adding a boolean column "is_actual_expense_reimbursement" to the ppm_shipments table in the db.  And, adding the boolean to GET and POST/PATCH swagger api's.

### Testing Reference
Swagger APIs:
* GET, 
  - Internal, GHC
    - GET /move/{locator}
      - Move Task Order ID:  response > root > id 
    - GET /move_task_orders/{moveTaskOrderID}/mto_shipments
      - isActualExpenseReimbursement: response > root > ppmShipment > isActualExpenseReimbursement
      - ppm shipment id: response > root > ppmShipment > id
* POST/PATCH
  - GHC
    - POST: /mto_shipments
    - PATCH: /move_task_orders/{moveId}/mto_shipments/{mtoShipmentId}
  - Internal, Prime, Prime v2, Prime v3
    - POST: /mto_shipments
    - PATCH: /mto_shipments/{mtoShipmentId}

SQL:
* Select row for PPM Shipment:
select *  from ppm_shipments ps where ps.id = 'ppm shipment id'

Swagger API URL:
* Internal: http://officelocal:3000/swagger-ui/internal.html#/
* GHC: http://officelocal:3000/swagger-ui/ghc.html#/
* Prime: http://officelocal:3000/swagger-ui/prime.html#/
* Prime v2: http://officelocal:3000/swagger-ui/prime_v2.html#/
* Prime v3: http://officelocal:3000/swagger-ui/prime_v3.html#/

### How to test
Prep:
* Create a move with a PPM shipment.

Swagger API: Get Actual Expense Reimbursement flag
* Internal API
  - login as customer
  - go to Internal API swagger page
  - get Move Task Order ID, run api: GET /move/{locator}
  - get isActualExpenseReimbursement, run api: GET /move_task_orders/{moveTaskOrderID}/mto_shipments
  - verify isActualExpenseReimbursement is null
  - update isActualExpenseReimbursement to true for the shipment in the database
  - run again: GET /move_task_orders/{moveTaskOrderID}/mto_shipments
  - verify isActualExpenseReimbursement is true
  - update isActualExpenseReimbursement to false for the shipment in the database
  - run again: GET /move_task_orders/{moveTaskOrderID}/mto_shipments
  - verify isActualExpenseReimbursement is false
* GHC API: 
  - login as counselor
  - rinse and repeat steps for Internal API
* Done

Swagger API: Post/Patch Actual Expense Reimbursement flag
* GHC
  - login as counselor
  - go to GHC Swagger API page
  - get Move Task Order ID, run api: GET /move/{locator}
  - POST /mto-shipments
    - Use template body for input below:
{
  "moveTaskOrderID": "1770f225-1da3-4f2e-9a71-3d793ba2ee05",
  "requestedPickupDate": "2024-10-18",
  "primeEstimatedWeight": 4500,
  "customerRemarks": "handle with care",
  "pickupAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "US",
    "county": "LOS ANGELES",
    "isOconus": false
  },
  "destinationAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "US",
    "county": "LOS ANGELES",
    "isOconus": false
  },
  "shipmentType": "PPM",
  "pointOfContact": "string",
  "counselorRemarks": "counselor approved",
"ppmShipment": {
  "pickupAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "US",
    "county": "LOS ANGELES",
    "isOconus": false
  },
  "destinationAddress": {
    "streetAddress1": "123 Main Ave",
    "streetAddress2": "Apartment 9000",
    "streetAddress3": "Montmârtre",
    "city": "Anytown",
    "state": "AL",
    "postalCode": "90210",
    "country": "US",
    "county": "LOS ANGELES",
    "isOconus": false
  },
    "actualDestinationPostalCode": null,
    "actualMoveDate": null,
    "actualPickupPostalCode": null,
    "advanceAmountReceived": null,
    "advanceAmountRequested": null,
    "approvedAt": null,
    "createdAt": "2024-10-18T16:12:11.495Z",
    "eTag": "MjAyNC0xMC0xOFQxNjoxMjoxMS40OTUxNzNa",
    "estimatedIncentive": 487082,
    "estimatedWeight": 4200,
    "expectedDepartureDate": "2024-10-18",
    "hasProGear": true,
    "hasReceivedAdvance": null,
    "hasRequestedAdvance": null,
    "id": "bff7400a-6a6f-44aa-892a-5249e02e36f4",
    "isActualExpenseReimbursement": true,
    "proGearWeight": 0,
    "reviewedAt": null,
    "shipmentId": "f2fd858a-8ee0-491e-abec-ab662b4c0d1a",
    "sitEstimatedCost": 25520,
    "sitEstimatedDepartureDate": "2024-10-18",
    "sitEstimatedEntryDate": "2024-10-18",
    "sitEstimatedWeight": 2000,
    "sitExpected": true,
    "sitLocation": "ORIGIN",
    "spouseProGearWeight": 0,
    "status": "SUBMITTED",
    "submittedAt": null,
    "updatedAt": "2024-10-18T16:12:11.495Z"
  }
}
    - Change "moveTaskOrderID" to the moveTaskOrderID from the GET
    - Verify in the response body that "isActualExpenseReimbursement" in PPM Shipment is true
    - Copy the following values from the POST response body
      - shipment id:  root > "id"
      - etag: root > "etag"
  - PATCH /move_task_orders/{moveTaskOrderID}/mto_shipments/{shipmentID}
    - inputs:
      - moveTaskOrderID: moveTaskOrderID from GET
      - shipmentID: shipment id from POST
      - etag: etag from POST
      - body: response body from POST
      - change "isActualExpenseReimbursement" in PPM Shipment to false
    - Execute PATCH
    - Verify in the response body that "isActualExpenseReimbursement" in PPM Shipment is false
* Internal, Prime, Prime v2, Prime v3
  - 'Rinse and repeat' GHC for each one.
    - Swagger webpages and api paths are slightly different, but instructions are the same.

## ANDI Screenshots
N/A

## Testing Screenshots
Database, ppm_shipments table, field is_actual_expense_reimbursement
![image](https://github.com/user-attachments/assets/af8bf186-509f-4862-9c3d-36e5404606f4)

Swagger api, GET, Internal and GHC
![image](https://github.com/user-attachments/assets/b0c10965-0640-48c5-a01f-f665e762b768)
![image](https://github.com/user-attachments/assets/ad0b24e3-09dc-4b36-8311-f071f26c1462)

Swagger, api, POST, GHC
![image](https://github.com/user-attachments/assets/fc511218-f6e3-46f4-ac5b-36283929cc8a)
![image](https://github.com/user-attachments/assets/62fa04c5-d471-4151-a393-2c499ff96ddb)

Swagger, api, PATCH, GHC
![image](https://github.com/user-attachments/assets/dd21327e-cf14-4742-8374-09ebb0cc6ef3)
![image](https://github.com/user-attachments/assets/2d46cba0-5e82-4299-b252-96f0a8b188a1)
